### DISCUSSION TITLE
Use the raw response in Reponse classes

### DISCUSSION THREAD

**Discussion by BinarSkugga** [fastapi/fastapi#6329] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class CustomEncoder():
    def encode(self, dict_data)
        return dict_data

class PhotonJSONResponse(JSONResponse):
    def __init__(self, content: typing.Any = None, status_code: int = 200, headers: dict = None, media_type: str = None,
                 background: BackgroundTask = None) -> None:
        # Fetch the untouched response in the upper stacks
        current_frame = inspect.currentframe()
        self.raw_response = None
        while current_frame.f_back:
            if 'raw_response' in current_frame.f_locals:
                self.raw_response = current_frame.f_locals['raw_response']
                break
            current_frame = current_frame.f_back
       
        self._encoder = CustomEncoder()
        super().__init__(content, status_code, headers, media_type, background)

    def render(self, content: Any) -> bytes:
        dict_data = self._encoder.encode(self.raw_response)
        return super().render(dict_data)
```

### Description

I want to access the raw response that hasn't been through the json_encoder inside my response class. This is because I have custom types that are handled in a custom encoder. I have looked through the relevant fastapi code and I can't find a way to override the encoder for all requests either. As you can see in the example code I currently use reflection to fetch the raw_response in the upper stack frame, however this is not very reliable. I also can't seem to do this using an APIRoute implementation because it would require re-implementing the route handler which is messy, maybe it would be more relevant in there though.

### Operating System

Windows

### FastAPI Version

0.63.0

### Python Version

3.8.12

**Comment by frankie567** [comment]
I think a better option here is to use the [`json_encoders` feature from Pydantic](https://pydantic-docs.helpmanual.io/usage/exporting_models/#json_encoders). It allows you to map a field type with a custom function to encode it.

You can see that FastAPI does ensure it applies them in `json_encodable` function:

https://github.com/tiangolo/fastapi/blob/1ce16c2f40a75a4d97233bf4e331ed563eda47cf/fastapi/encoders.py#L52-L55

**Comment by BinarSkugga** [comment]
Except my types are not instances of BaseModel. For example, namedtuple are expressed as a list in json while I want a dictionary so I don't lose the keys information.

**Comment by frankie567** [comment]
Well, I think you are on a very specific use-case here. My primary advise would be to first have proper Pydantic models counterpart for your internal data types, so you get all the benefits of FastAPI.

If you don't want to go that way, why not return the response object yourself in the endpoint?

```py
from collections import namedtuple

Point = namedtuple('Point', ['x', 'y'])  # Dummy type

class CustomEncoderJSONResponse(JSONResponse):
    def render(self, content: Any) -> bytes:
        encoder = CustomEncoder()
        # At this point, content is still whatever you passed it into
        dict_data = encoder.encode(content)
        return super().render(dict_data)

@app.get("/point")
async def get_point():
    p = Point(x=3, y=2)
    return CustomEncoderJSONResponse(content=p)
```

**Comment by BinarSkugga** [comment]
First, thanks for the help. I'm not a fan of boxing in general so I don't think I'll go the route of wrapping my tuples into `BaseModel`s. I was kind of liking the idea of simply returning the `namedtuple` without the overhead of creating a response too. I could make a `BaseModel` that behaves like a `namedtuple` by using the internal `__dict__` and specifying attributes at runtime but this also feels a bit hacky. I currently have a working implementation of the `APIRoute` that bypasses the `json_encoder` based on a parameter. It feels a bit hacky because it goes a bit against the fastapi philosophy but its going to do for now. Thanks again !
